[JSC] LiteralParser Lexer should not do double-lookup for dispatch

Constellation opened the pull request:
#### babf286f913efa4e6223d73bdbe3d720ad28f430
<pre>
[JSC] LiteralParser Lexer should not do double-lookup for dispatch
<a href="https://bugs.webkit.org/show_bug.cgi?id=296915">https://bugs.webkit.org/show_bug.cgi?id=296915</a>
<a href="https://rdar.apple.com/157535403">rdar://157535403</a>

Reviewed by Keith Miller.

We do the same optimization to #48939, for LiteralParser. We do not need
to have double-lookup for dispatching lexing. We should just have a
character and dispatch a switch directly with that.

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::isJSONWhiteSpace):
(JSC::reviverMode&gt;::Lexer::lex):

Canonical link: <a href="https://commits.webkit.org/298258@main">https://commits.webkit.org/298258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d5450e081cca4ce5b72b20384bafaef9da46f18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114843 "100 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34585 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120998 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65525 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/91f29529-21e9-4a1d-b78a-5e37cd902751) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35215 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43144 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87284 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c26ea47e-6f54-43fb-bc48-347bd981efcb) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117791 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103124 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67676 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27241 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21245 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64661 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107181 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97444 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21358 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124196 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113395 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31267 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42220 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95875 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41061 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18901 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18388 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41725 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47241 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137610 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41282 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36793 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44597 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43025 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -729,7 +729,10 @@ static constexpr const bool safeStringLatin1CharactersInStrictJSON[256] = {
 static ALWAYS_INLINE bool isJSONWhiteSpace(const CharType& c)
 {
-    return tokenTypesOfLatin1Characters[static_cast<uint8_t>(c)] == TokErrorSpace && isLatin1(c);
+    if constexpr (sizeof(CharType) == sizeof(LChar))
```
Nit: I would do `std::is_same_v<CharType, LChar>`

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
-            }
-            auto result = lexString<hint>(token, character);
+    LChar type = character;
+    if constexpr (sizeof(CharType) != sizeof(LChar)) {
```
Ditto.

Constellation replied:
Done

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
+    }
+
+    switch (type) {
+    case  39 /* '                  TokString */: {
```
Can we replace 39 with `'\''`? There also doesn't need to be so many spaces in this comment.

Constellation replied:
I feel like 39 is better since it is ensuring all the numbers are listed in this switch. `'` is already noted in the comment, so I think this is more informative.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
         }
+        [[fallthrough]];
+    }
+    case  34 /* "                  TokString */: {
```
Ditto `'"'` and spaces.

Constellation replied:
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
-                break;
-            }
-            auto result = lexIdentifier(token);
+    case 116 /* t                  TokIdentifier */: {
```
Ditto `t` and spaces

Constellation replied:
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
 
-        case TokNumber: {
-            auto result = lexNumber(token);
+    case 102 /* f                  TokIdentifier */: {
```
Ditto 'f' and spaces.

Constellation replied:
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
-                || tokenType == TokSemi);
-            token.type = tokenType;
-            ++m_ptr;
+    case 110 /* n                  TokIdentifier */: {
```
ditto.

Constellation replied:
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
-    if constexpr (reviverMode == JSONReviverMode::Enabled)
-        m_currentTokenEnd = m_ptr;
+
+    case  36 /* $                  TokIdentifier */:
```
Can we run `/(\d+)\ \/\* (.)\s+(\w+) /'$2' /* $3 /s` here? At least for the ones that have latin letters.

Constellation replied:
I'm intentionally using the number to ensure that all the numbers 0-255 are listed in this switch, and I think that using the letter makes it harder to see this. So I prefer to the current style, and it is also having a comment describing the content, so it is also more informative than using the letter.

kmiller68 reviewed (approved):
r=me with nits.

kmiller68 reviewed (approved):
r=me with nits. Now that I wrote the comments but saw the non-ascii characters I'm a bit less worried about the `case (\d+)` replacements.

kmiller68 reviewed (approved):
r=me with nits. Now that I wrote the comments but saw the non-ascii characters I'm a bit less worried about the `case (\d+)` replacements.

kmiller68 commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -756,91 +759,349 @@ ALWAYS_INLINE TokenType LiteralParser<CharType, reviverMode>::Lexer::lex(Literal
-            }
-            auto result = lexIdentifier(token);
+    case 116 /* t                  TokIdentifier */: {
+        if (m_end - m_ptr >= 4 && compareCharacters(m_ptr + 1, 'r', 'u', 'e')) [[likely]] {
```
Does `compareCharacters` compile to a single load? Maybe worth taking a look at that too.

Constellation replied:
Yeah, I think `compareCharacters(m_ptr, 't', 'r', 'u', 'e')` will be better. Changed.

kmiller68 reviewed (approved):
r=me with nits. Now that I wrote the comments but saw the non-ascii characters I'm a bit less worried about the `case (\d+)` replacements.

Constellation committed:
[JSC] LiteralParser Lexer should not do double-lookup for dispatch
https://bugs.webkit.org/show_bug.cgi?id=296915
rdar://157535403

Reviewed by Keith Miller.

We do the same optimization to #48939, for LiteralParser. We do not need
to have double-lookup for dispatching lexing. We should just have a
character and dispatch a switch directly with that.

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::isJSONWhiteSpace):
(JSC::reviverMode>::Lexer::lex):

Canonical link: https://commits.webkit.org/298258@main
